src/index.js

```diff
@@ -45,30 +56,30 @@ export default function urlJoin(...parts) {
     .filter((part) => typeof part === 'string' || typeof part === 'number')
     .join('/');
 
-    // Split the parts into beforePathname, pathname, and afterPathname
+    // Split the parts into prefix, pathname, and suffix
     // (scheme://host)(/pathname)(?queryString)
-    const { beforePathname, pathname, afterPathname } = splitUrl(partsStr, options);
+    const { prefix, pathname, suffix } = parseUrl(partsStr, options);
```

Comment:
The code below needs now to take into consideration the `pathname.hasLeadingSlash`, `pathname.hasTrailingSlash`, `options.leadingSlash` and `options.trailingSlash` to correctly build the final url.

Reply:
I've just added more code to deal with the options `leadingSlash` and `trailingSlash`, but I'd like to check one thing:

For `/foo/bar/` with `options.trailingSlash = 'keep'` and `options.leadingSlash = 'keep'`, the correct output of the `parseUrl` function would be this one below?
```js
 {
   prefix: '',
   pathname: {
     parts: [ '/foo', 'bar/' ],
     hasLeading: true,
     hasTrailing: true
  },
  suffix: ''
}
```

Reply:
`parseUrl` should always contain no slashes in the parts:

 ```js
 {
   prefix: '',
   pathname: {
     parts: [ 'foo', 'bar' ],
     hasLeading: true,
     hasTrailing: true
  },
  suffix: ''
}
```

The concatenation should be done in the `urlJoin` function or perhaps into a new `recreateUrl` / `rebuildUrl` function. Because it knows that it should preserve both trailing and leading, it would respect what `pathname.hasLeading` and `pathname.hasTrailing` says. In this case, it would recreate the following URL: `/foo/bar/`.

Reply:
But, as an example, if `options.leadingSlash` was true and `options.trailingSlash` was false, it would recreate the following URL: `/foo/bar` (it would ignore what `pathname.hasTrailing` says).

Reply:
Oh ok! My bad, I've just commited the wrong way of doing it. I'll undo this last commit and start working on this `recreateUrl` / `rebuildUrl`. Thanks for the info

Reply:
No problem! If we were things personally, these misunderstandings wouldn't happen 👍

Reply:
Could you please confirm me which are the expected outputs for the following cases?

1. When `options = { trailingSlash: true }`, what `urlJoin(options)` should return?
2. When `options = { trailingSlash:  'keep' }`, what `urlJoin(options)` should return?
3. When `options = { leadingSlash:  'keep' }`, what `urlJoin(options)` should return?

Reply:
1. `/`
2. ``
3. ``

Reply:
>     1. `/`
> 
>     2. `<empty>`
> 
>     3. `<empty>`

Got it! Thanks

Reply:
> Could you please confirm me which are the expected outputs for the following cases?
> 
>     1. When `options = { trailingSlash: true }`, what `urlJoin(options)` should return?
> 
>     2. When `options = { trailingSlash:  'keep' }`, what `urlJoin(options)` should return?
> 
>     3. When `options = { leadingSlash:  'keep' }`, what `urlJoin(options)` should return?

I was able to address the first case, but I'm stuck on the cases 2 and 3.

Case 2: the problem here is that at line 54 the leading slash is being added and it simply remains there, so the final output will be `/` instead of `<empty>`. I thought about adding another condition on the `addLeadingSlash` function, so it would leave this context without the slash, but it would still be added by the `addTrailingSlash` later on the pipeline. I'm not sure which is the better approach here, where I should step in.

Case 3: the problem here is in the way I created the `addTrailingSlash` function it will add the slash if the current url does not end with `/`.

Any suggestion on how should I fix those?

Reply:
See my comment below on the PR

Reply:
Oh ok, nevermind then! Thanks!